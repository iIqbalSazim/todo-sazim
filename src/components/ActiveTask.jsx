import { ActionIcon, Badge, Card, Code, Group, Text } from "@mantine/core";
import {
  IconCircleCheck,
  IconCircleMinus,
  IconEditCircle,
} from "@tabler/icons-react";

const ActiveTask = ({
  task,
  assignColorByPriority,
  toggleIsCompleted,
  openEditForm,
  deleteTask,
}) => {
  return (
    <Card
      bg={assignColorByPriority(task.priority)}
      shadow={"md"}
      radius={"md"}
      px="xl"
      py="xs"
      mx="xl"
      withBorder
    >
      <Group position="apart">
        <Text weight={500} size={"15px"} maw={"70%"}>
          {task.description}
        </Text>
        <Group position="right">
          <Badge color={assignColorByPriority(task.priority)} variant="light">
            {task.priority}
          </Badge>
          <ActionIcon
            variant="light"
            onClick={() => toggleIsCompleted(task.id)}
            bg="teal.1"
          >
            <IconCircleCheck size="1.125rem" />
          </ActionIcon>
          <ActionIcon
            variant="light"
            onClick={() => openEditForm(task)}
            bg="white"
          >
            <IconEditCircle size="1.125rem" />
          </ActionIcon>
          <ActionIcon
            variant="light"
            bg="red.1"
            onClick={() => deleteTask(task.id)}
          >
            <IconCircleMinus size="1.125rem" />
          </ActionIcon>
        </Group>
      </Group>
      <Group position="apart" pt={"md"}>
        <Text size="xs" weight={300}>
          Created at {task.createdAt}
        </Text>
        <Code size={"sm"} radius={"md"} color="red">
          Complete by: {task.dueDate}
        </Code>
      </Group>
    </Card>
  );
};

export default ActiveTask;
