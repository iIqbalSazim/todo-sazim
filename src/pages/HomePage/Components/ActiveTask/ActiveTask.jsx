import {
  ActionIcon,
  Badge,
  Card,
  Code,
  Group,
  MediaQuery,
  Text,
} from "@mantine/core";
import {
  IconCircleCheck,
  IconCircleMinus,
  IconEditCircle,
} from "@tabler/icons-react";

import { formatCompletedAt, formatDueDate } from "../../HomePageHelpers";

const ActiveTask = ({
  task,
  assignColorByPriority,
  toggleIsCompleted,
  openEditForm,
  archiveTask,
}) => {
  return (
    <>
      <MediaQuery smallerThan={"sm"} styles={{ display: "none" }}>
        <Card
          bg={assignColorByPriority(task.priority)}
          shadow={"md"}
          radius={"md"}
          px="xl"
          py="xs"
          mx="xl"
          withBorder
        >
          <Group position="apart">
            <Text weight={500} size={"lg"} maw={"70%"}>
              {task.description}
            </Text>
            <Group position="right">
              <Badge
                color={assignColorByPriority(task.priority)}
                variant="light"
              >
                {task.priority}
              </Badge>
              <ActionIcon
                variant="light"
                onClick={() => toggleIsCompleted(task.id)}
                bg="teal.1"
              >
                <IconCircleCheck
                  data-testid="toggle-complete"
                  size="1.125rem"
                />
              </ActionIcon>
              <ActionIcon
                variant="light"
                onClick={() => openEditForm(task)}
                bg="white"
              >
                <IconEditCircle data-testid="edit" size="1.125rem" />
              </ActionIcon>
              <ActionIcon
                variant="light"
                bg="red.1"
                onClick={() => archiveTask(task)}
              >
                <IconCircleMinus data-testid="archive" size="1.125rem" />
              </ActionIcon>
            </Group>
          </Group>
          <Group position="apart" pt={"md"}>
            <Text size="xs" weight={300}>
              Created at {formatCompletedAt(task.created_at)}
            </Text>
            <Code
              size={"sm"}
              radius={"md"}
              bg={assignColorByPriority(task.priority)}
            >
              Complete by: {formatDueDate(task.due_date)}
            </Code>
          </Group>
        </Card>
      </MediaQuery>

      <MediaQuery largerThan={"sm"} styles={{ display: "none" }}>
        <Card
          bg={assignColorByPriority(task.priority)}
          shadow={"sm"}
          radius={"md"}
          py="sm"
          withBorder
        >
          <Group position="right" spacing={"sm"}>
            <Badge
              color={assignColorByPriority(task.priority)}
              variant="light"
              size="sm"
            >
              {task.priority}
            </Badge>
            <ActionIcon
              variant="light"
              onClick={() => toggleIsCompleted(task.id)}
              bg="teal.1"
            >
              <IconCircleCheck size="1.2rem" />
            </ActionIcon>
            <ActionIcon
              variant="light"
              onClick={() => openEditForm(task)}
              bg="white"
            >
              <IconEditCircle size="1.2rem" />
            </ActionIcon>
            <ActionIcon
              variant="light"
              bg="red.1"
              onClick={() => archiveTask(task)}
            >
              <IconCircleMinus size="1.2rem" />
            </ActionIcon>
          </Group>
          <Text weight={500} size={"lg"} w={"100%"} py={"xs"}>
            {task.description}
          </Text>
          <Group position="right">
            <Code
              size={"xs"}
              radius={"md"}
              bg={assignColorByPriority(task.priority)}
            >
              Complete by: {formatDueDate(task.due_date).slice(4)}
            </Code>
          </Group>
        </Card>
      </MediaQuery>
    </>
  );
};

export default ActiveTask;
