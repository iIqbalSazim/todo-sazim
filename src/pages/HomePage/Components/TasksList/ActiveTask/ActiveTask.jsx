import {
  ActionIcon,
  Badge,
  Card,
  Code,
  Group,
  MediaQuery,
  Text,
} from "@mantine/core";
import {
  IconCircleCheck,
  IconCircleMinus,
  IconEditCircle,
} from "@tabler/icons-react";

const ActiveTask = ({
  task,
  assignColorByPriority,
  toggleIsCompleted,
  openEditForm,
  deleteTask,
}) => {
  return (
    <>
      <MediaQuery smallerThan={"sm"} styles={{ display: "none" }}>
        <Card
          bg={assignColorByPriority(task.priority)}
          shadow={"md"}
          radius={"md"}
          px="xl"
          py="xs"
          mx="xl"
          withBorder
        >
          <Group position="apart">
            <Text weight={500} size={"lg"} maw={"70%"}>
              {task.description}
            </Text>
            <Group position="right">
              <Badge
                color={assignColorByPriority(task.priority)}
                variant="light"
              >
                {task.priority}
              </Badge>
              <ActionIcon
                variant="light"
                onClick={() => toggleIsCompleted(task.id)}
                bg="teal.1"
              >
                <IconCircleCheck size="1.125rem" />
              </ActionIcon>
              <ActionIcon
                variant="light"
                onClick={() => openEditForm(task)}
                bg="white"
              >
                <IconEditCircle size="1.125rem" />
              </ActionIcon>
              <ActionIcon
                variant="light"
                bg="red.1"
                onClick={() => deleteTask(task.id)}
              >
                <IconCircleMinus size="1.125rem" />
              </ActionIcon>
            </Group>
          </Group>
          <Group position="apart" pt={"md"}>
            <Text size="xs" weight={300}>
              Created at {task.created_at}
            </Text>
            <Code
              size={"sm"}
              radius={"md"}
              bg={assignColorByPriority(task.priority)}
            >
              Complete by: {task.due_date}
            </Code>
          </Group>
        </Card>
      </MediaQuery>

      <MediaQuery largerThan={"sm"} styles={{ display: "none" }}>
        <Card
          bg={assignColorByPriority(task.priority)}
          shadow={"sm"}
          radius={"md"}
          py="sm"
          withBorder
        >
          <Group position="right" spacing={"sm"}>
            <Badge
              color={assignColorByPriority(task.priority)}
              variant="light"
              size="sm"
            >
              {task.priority}
            </Badge>
            <ActionIcon
              variant="light"
              onClick={() => toggleIsCompleted(task.id)}
              bg="teal.1"
            >
              <IconCircleCheck size="1.2rem" />
            </ActionIcon>
            <ActionIcon
              variant="light"
              onClick={() => openEditForm(task)}
              bg="white"
            >
              <IconEditCircle size="1.2rem" />
            </ActionIcon>
            <ActionIcon
              variant="light"
              bg="red.1"
              onClick={() => deleteTask(task.id)}
            >
              <IconCircleMinus size="1.2rem" />
            </ActionIcon>
          </Group>
          <Text weight={500} size={"lg"} w={"100%"} py={"xs"}>
            {task.description}
          </Text>
          <Group position="right">
            <Code
              size={"xs"}
              radius={"md"}
              bg={assignColorByPriority(task.priority)}
            >
              Complete by:{" "}
              {task.due_data ? task.due_date.slice(4) : "no due date"}
            </Code>
          </Group>
        </Card>
      </MediaQuery>
    </>
  );
};

export default ActiveTask;
