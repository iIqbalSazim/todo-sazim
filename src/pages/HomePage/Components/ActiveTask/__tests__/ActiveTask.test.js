import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import ActiveTask from "../ActiveTask";

const mockToggleIsCompleted = jest.fn();
const mockOpenEditForm = jest.fn();
const mockArchiveTask = jest.fn();

const mockTask = {
  id: 1,
  description: "This is a test",
  priority: "high",
  created_at: new Date(2023, 11, 2, 14, 17, 0, 0),
  due_date: "Sat Dec 02 2023",
};

describe("ActiveTask", () => {
  test("renders correctly", () => {
    const { container } = render(
      <ActiveTask
        task={mockTask}
        assignColorByPriority={() => {}}
        toggleIsCompleted={mockToggleIsCompleted}
        openEditForm={mockOpenEditForm}
        archiveTask={mockArchiveTask}
      />
    );

    const taskDescriptionElements = screen.getAllByText(mockTask.description);
    const taskPriorityElements = screen.getAllByText(mockTask.priority);
    const createdAtElement = screen.getByText(/created at/i);
    const dueDateElements = screen.getAllByText(/complete by/i);
    const toggleCompleteButton = screen.getByTestId("toggle-complete");
    const openEditFormButton = screen.getByTestId("edit");
    const archiveButton = screen.getByTestId("archive");

    expect(taskDescriptionElements.length).toBe(2);
    expect(taskPriorityElements.length).toBe(2);
    expect(createdAtElement).toBeInTheDocument();
    expect(dueDateElements.length).toBe(2);
    expect(toggleCompleteButton).toBeInTheDocument();
    expect(openEditFormButton).toBeInTheDocument();
    expect(archiveButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("calls toggleIsCompleted with task id as argument on check-icon button click", async () => {
    render(
      <ActiveTask
        task={mockTask}
        assignColorByPriority={() => {}}
        toggleIsCompleted={mockToggleIsCompleted}
        openEditForm={mockOpenEditForm}
        archiveTask={mockArchiveTask}
      />
    );

    const toggleCompleteButton = screen.getByTestId("toggle-complete");

    await userEvent.click(toggleCompleteButton);

    expect(mockToggleIsCompleted).toHaveBeenCalledWith(mockTask.id);
  });

  test("calls openEditForm with task as argument on edit-icon button click", async () => {
    render(
      <ActiveTask
        task={mockTask}
        assignColorByPriority={() => {}}
        toggleIsCompleted={mockToggleIsCompleted}
        openEditForm={mockOpenEditForm}
        archiveTask={mockArchiveTask}
      />
    );

    const openEditFormButton = screen.getByTestId("edit");

    await userEvent.click(openEditFormButton);

    expect(mockOpenEditForm).toHaveBeenCalledWith(mockTask);
  });

  test("calls archiveTask with task as argument on minus-icon button click", async () => {
    render(
      <ActiveTask
        task={mockTask}
        assignColorByPriority={() => {}}
        toggleIsCompleted={mockToggleIsCompleted}
        openEditForm={mockOpenEditForm}
        archiveTask={mockArchiveTask}
      />
    );

    const archiveButton = screen.getByTestId("archive");

    await userEvent.click(archiveButton);

    expect(mockArchiveTask).toHaveBeenCalledWith(mockTask);
  });
});
