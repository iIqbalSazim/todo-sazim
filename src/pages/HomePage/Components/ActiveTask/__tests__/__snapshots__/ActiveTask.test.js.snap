// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveTask renders correctly 1`] = `
<div>
  <div
    class="mantine-Paper-root mantine-Card-root mantine-MediaQuery-media mantine-6hmcmv"
    data-with-border="true"
  >
    <div
      class="mantine-Group-root mantine-1emc9ft"
    >
      <div
        class="mantine-Text-root mantine-3efcew"
      >
        This is a test
      </div>
      <div
        class="mantine-Group-root mantine-kivjf7"
      >
        <div
          class="mantine-Badge-root mantine-4y3x5h"
        >
          <span
            class="mantine-1t45alw mantine-Badge-inner"
          >
            high
          </span>
        </div>
        <button
          class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-1u9dylz"
          type="button"
        >
          <svg
            class="tabler-icon tabler-icon-circle-check"
            data-testid="toggle-complete"
            fill="none"
            height="1.125rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.125rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
            />
            <path
              d="M9 12l2 2l4 -4"
            />
          </svg>
        </button>
        <button
          class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-1r41os4"
          type="button"
        >
          <svg
            class="tabler-icon tabler-icon-edit-circle"
            data-testid="edit"
            fill="none"
            height="1.125rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.125rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 15l8.385 -8.415a2.1 2.1 0 0 0 -2.97 -2.97l-8.415 8.385v3h3z"
            />
            <path
              d="M16 5l3 3"
            />
            <path
              d="M9 7.07a7 7 0 0 0 1 13.93a7 7 0 0 0 6.929 -6"
            />
          </svg>
        </button>
        <button
          class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-43sb5"
          type="button"
        >
          <svg
            class="tabler-icon tabler-icon-circle-minus"
            data-testid="archive"
            fill="none"
            height="1.125rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.125rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
            />
            <path
              d="M9 12l6 0"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="mantine-Group-root mantine-vxpk8m"
    >
      <div
        class="mantine-Text-root mantine-1xp5ood"
      >
        Created at 
        2:17 p.m. | Sat Dec 02 2023
      </div>
      <code
        class="mantine-Code-root mantine-fkl7uc"
        dir="ltr"
        radius="md"
      >
        Complete by: 
        Sat Dec 02 2023
      </code>
    </div>
  </div>
  <div
    class="mantine-Paper-root mantine-Card-root mantine-MediaQuery-media mantine-1qgaa8u"
    data-with-border="true"
  >
    <div
      class="mantine-Group-root mantine-1v5jpi4"
    >
      <div
        class="mantine-Badge-root mantine-dd77nk"
      >
        <span
          class="mantine-1t45alw mantine-Badge-inner"
        >
          high
        </span>
      </div>
      <button
        class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-1u9dylz"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-circle-check"
          fill="none"
          height="1.2rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.2rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
          />
          <path
            d="M9 12l2 2l4 -4"
          />
        </svg>
      </button>
      <button
        class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-1r41os4"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-edit-circle"
          fill="none"
          height="1.2rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.2rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 15l8.385 -8.415a2.1 2.1 0 0 0 -2.97 -2.97l-8.415 8.385v3h3z"
          />
          <path
            d="M16 5l3 3"
          />
          <path
            d="M9 7.07a7 7 0 0 0 1 13.93a7 7 0 0 0 6.929 -6"
          />
        </svg>
      </button>
      <button
        class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-43sb5"
        type="button"
      >
        <svg
          class="tabler-icon tabler-icon-circle-minus"
          fill="none"
          height="1.2rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.2rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
          />
          <path
            d="M9 12l6 0"
          />
        </svg>
      </button>
    </div>
    <div
      class="mantine-Text-root mantine-xkqxxs"
    >
      This is a test
    </div>
    <div
      class="mantine-Group-root mantine-kivjf7"
    >
      <code
        class="mantine-Code-root mantine-fkl7uc"
        dir="ltr"
        radius="md"
      >
        Complete by: 
        Dec 02 2023
      </code>
    </div>
  </div>
</div>
`;
